### ISSUE TITLE
keras Callbacks without _supports_tf_logs update separate `logs` 

### ISSUE THREAD

**Issue by jackd** [tensorflow/tensorflow#45895]
## System information
- Have I written custom code: yes
- OS Platform and Distribution: Ubuntu 18.04:
- TensorFlow installed from: binary
- TensorFlow version: v1.12.1-47912-gec43aacb56f 2.5.0-dev20201219
- Python version: 3.7.7

## Current Behaviour
Callbacks in `CallbackList` update different `logs` dicts based on private `_supports_tf_logs` attribute (see e.g. [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/callbacks.py#L431)). This attribute is undocumented and it is unclear exactly what it is intended to signify. If a callback adds an entry to `logs`, the effect this has depends on this value.

## Expected behavior
Callbacks should be affected to previous callbacks' mutations of logs regardless of `_supports_tf_logs` values.

## Standalone code
[This colab](https://colab.research.google.com/drive/1Gd6JaZZk9fKmZSFdX7yd9Q_Oui-YzqDv?usp=sharing) shows the result of manually changing the property value of `LearningRateScheduler` and it's affect on `ProgbarLogger`'s behaviour.

Code reproduced below for convenience
```python
import tensorflow as tf
inp = tf.keras.Input((1,))
out = tf.keras.layers.Dense(1)(inp)
model = tf.keras.Model(inp, out)
model.compile(loss='mse', optimizer='sgd')

x = tf.random.uniform((10, 1))
y = 2 * x + 3
dataset = tf.data.Dataset.from_tensor_slices((x, y)).batch(2)

sched = tf.keras.callbacks.LearningRateScheduler(lambda i: 1. / (i+1))
sched._supports_tf_logs = True  # makes ProgbarLogger display lr
# same issue with ReduceLROnPlateau
callbacks = [sched]

# add logger at end, otherwise it's inserted at front and won't print lr
callbacks.append(tf.keras.callbacks.ProgbarLogger())

model.fit(dataset, epochs=10, callbacks=callbacks)
```
Output
```txt
Epoch 1/10
5/5 [==============================] - 1s 109ms/sample - loss: 58.7811 - lr: 1.0000
Epoch 2/10
5/5 [==============================] - 0s 2ms/sample - loss: 94.1334 - lr: 0.5000
Epoch 3/10
5/5 [==============================] - 0s 2ms/sample - loss: 22.3592 - lr: 0.3333
Epoch 4/10
5/5 [==============================] - 0s 1ms/sample - loss: 3.3667 - lr: 0.2500
Epoch 5/10
5/5 [==============================] - 0s 1ms/sample - loss: 1.4383 - lr: 0.2000
Epoch 6/10
5/5 [==============================] - 0s 1ms/sample - loss: 0.8452 - lr: 0.1667
Epoch 7/10
5/5 [==============================] - 0s 1ms/sample - loss: 0.5714 - lr: 0.1429
Epoch 8/10
5/5 [==============================] - 0s 1ms/sample - loss: 0.4179 - lr: 0.1250
Epoch 9/10
5/5 [==============================] - 0s 1ms/sample - loss: 0.3216 - lr: 0.1111
Epoch 10/10
5/5 [==============================] - 0s 1ms/sample - loss: 0.2565 - lr: 0.1000
```

**Comment by rmothukuru** [comment]
@jackd,
Setting **`sched._supports_tf_logs = True`** prints the Value of `Learning Rate` in the `Logs` and setting it to `False` doesn't prints its value. That is the difference I observe and the name `supports_tf_logs` seems apt to me. 

Please find the [Colab Gist](https://colab.research.google.com/gist/rmothukuru/58719147c8113148e00b15acec8d40b6/callback-logs.ipynb).

If this is not what you mean, can you please demonstrate the behavior in your code? Thanks!

**Comment by jackd** [comment]
@rmothukuru there is a very significant difference between the names `supports_tf_logs` and `_supports_tf_logs`.

The current behaviour is:

- the first callback with `_supports_tf_logs = False` observes mutations to `logs` only from prior callbacks
- subsequent callbacks with `_supports_tf_logs = False` observe mutations to `logs` only from callbacks up to the first callback with `_supports_tf_logs=False` and those with `_supports_tf_logs=False`
- callbacks with `_supports_tf_logs=True` observe mutations only from previous callbacks with `_supports_tf_logs=True`

Public/protected attribute aside, the name `supports_tf_logs` does not imply *any* of those to me. In the example provided, why does `ProgbarLogger`'s behaviour change based on whether `LearningRateScheduler` supports tf logs?

**Comment by rmothukuru** [comment]
@jackd,
What I meant was **`_supports_tf_logs`**, not **`supports_tf_logs`**. Sorry for the confusion.

I didn't observe change in `ProgbarLogger`'s behavior in [this Gist](https://colab.research.google.com/gist/rmothukuru/58719147c8113148e00b15acec8d40b6/callback-logs.ipynb#scrollTo=7k7wwszKEAB1) based on the value of **`_supports_tf_logs`**. Can you please elaborate about it.

Thanks!

**Comment by jackd** [comment]
Cell 2's output has `lr`, Cell 3's doesn't.

**Comment by rmothukuru** [comment]
@jackd,
Is that not expected because it is changing as per the value of 
**`tf.keras.callbacks.LearningRateScheduler(lambda i: 1. / (i+1))._supports_tf_logs`**?

**Comment by jackd** [comment]
... I suppose that depends what you mean by "expect". I expect it because I'm very familiar with the implementation, because I've gone digging when `lr` disappeared from my logs after a tf update, in the same way that I expect any bug to remain there until it gets fixed. I certainly didn't expect it before I went digging. It's not part of the public API, nor is it documented anyway, nor do I believe this partitioning "feature" of callbacks was intended by whoever changed things.

**Comment by shi510** [comment]
This issue should be solved.
I have a custom callback that evaluates recall metric on epoch end.
```python
class RecallCallback(tf.keras.callbacks.Callback):
  def on_epoch_end(self, epoch, logs=None):
      logs['recall@1'] = evaluate_recall(...)
```
```python
checkpoint = ModelCheckpoint(monitor='recall@1', ...)
#checkpoint._supports_tf_logs = False
net.fit(callbacks=[RecallCallback(), checkpoint], ...)
```
It does not sync the updates of logs.
But it works by setting the _supports_tf_logs to False.

**Comment by lgeiger** [comment]
@omalleyt12 @fchollet Could you take a look at this issue? This seems to break many existing callbacks, so it would be great if this could be fixed, or documented clearly that `logs`  shouldn't be mutated.

**Comment by lgeiger** [comment]
I made a PR to fix this: https://github.com/tensorflow/tensorflow/pull/47922
